### PR TITLE
fix(deps): updated vulnerable dependencies

### LINKED ISSUE #1532: Improve Scorecard Score for Kubearmor
## Feature Request

**Short Description**

Improve the aggregate [OSSF/Scorecard](https://securityscorecards.dev/viewer/?uri=github.com/kubearmor/KubeArmor) score received for Kubearmor
Currently, the score we received was **5.1/10**.

**Is your feature request related to a problem? Please describe the use case.**

No

**Describe the solution you'd like**

Work on each area to analyze where the score is dropped and how we can improve upon it!

- [X] Dangerous Workflow
- [x]  Binary Artifacts
- [x]  Branch Protection
- [X]  CI Tests
- [x]  Code Review
- [X]  Contributors
- [X]  Dependency-Update-Tool
- [ ]  Fuzzing
- [X]  Maintained
- [X]  Packaging
- [ ]  SAST
- [X]  Security Policy
- [ ]  Signed-Releases
- [x] Vulnerabilities
- [ ] Token-Permission

<!DOCTYPE html><h2 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.5em; font-weight: var(--base-text-weight-semibold, 600); line-height: 1.25; padding-bottom: 0.3em; border-bottom: 1px solid var(--borderColor-muted, var(--color-border-muted)); caret-color: rgb(230, 237, 243); color: rgb(230, 237, 243); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-caps: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"><strong style="box-sizing: border-box; font-weight: var(--base-text-weight-semibold, 600);">Scorecard Result Details</strong></h2><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; caret-color: rgb(230, 237, 243); color: rgb(230, 237, 243); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"><strong style="box-sizing: border-box; font-weight: var(--base-text-weight-semibold, 600);">Aggregate score: 5.1 / 10</strong></p>
SCORE | NAME | REASON | DOCUMENTATION/REMEDIATION
-- | -- | -- | --
10 / 10 | Dangerous-Workflow | No dangerous workflow detected | https://github.com/ossf/scorecard/blob/4edb07802fdad892fa8d10f8fd47666b6ccc27c9/docs/checks.md#dangerous-workflow
0 / 10 | Binary-Artifacts | multiple binary artifacts found | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#binary-artifacts
2 / 10 | Branch-Protection | branch protection is not maximal on main and all release branches | https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule and https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule
10 / 10 | CI-Tests | 30 out of 30 merged PRs checked by a CI test -- score normalized to 10 | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#ci-tests
5 / 10 | CII-Best-Practices | badge detected: passing | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#cii-best-practices
10 / 10 | Code-Review | GitHub code reviews found for 30 commits out of the last 30 -- score normalized to 10 | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#code-review
10 / 10 | Contributors | 44 different companies found -- score normalized to 10 | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#contributors
10 / 10 | Dependency-Update-Tool | update tool detected | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#dependency-update-tool
0 / 10 | Fuzzing | project is not fuzzed | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#fuzzing
10 / 10 | Maintained | 30 commit(s) out of 30 and 30 issue activity out of 30 found in the last 90 days -- score normalized to 10 | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#maintained
10 / 10 | Packaging | publishing workflow detected | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#packaging
0 / 10 | Pinned-Dependencies | No dependencies are pinned | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#pinned-dependencies
0 / 10 | SAST | SAST tool is not run on all commits -- score normalized to 0 | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#sast and https://docs.github.com/en/code-security/code-scanning/enabling-code-scanning/configuring-default-setup-for-code-scanning#configuring-default-setup-for-a-repository
10 / 10 | Security-Policy | security policy file detected | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#security-policy
0 / 10 | Signed-Releases | 0 out of 5 artifacts are signed -- score normalized to 0 | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#signed-releases and https://wiki.debian.org/Creating%20signed%20GitHub%20releases
0 / 10 | Token-Permissions | non read-only tokens detected in GitHub workflows, tokens are not managed | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#token-permissions
10 / 10 | Vulnerabilities | **Several vulnerabilities detected** | https://github.com/ossf/scorecard/blob/23b0ddb8aa96356321cf31a2709723e29b15a951/docs/checks.md#vulnerabilities

CC @daemon1024 @nyrahul

### PR DESCRIPTION
**Pull request by rootxrishabh**
**Purpose of PR?**:
This PR updates all the vulnerable dependencies alerted by the OSSF scorecard.
Fixes #
#1532 
**Does this PR introduce a breaking change?**
No
**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Additional information for reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [X] Bug fix. Fixes #<issue number>
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [X] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

**Comment by DelusionalOptimist** [comment]
@rootxrishabh can you try by removing these k8s.io replace directives across KubeArmor's and other go.mod files? I think it should fix the current build failure. :thinking: 
https://github.com/kubearmor/KubeArmor/blob/2142379067947bd1b892dd48f436f9951b666d18/KubeArmor/go.mod#L22-L24

**Comment by rootxrishabh** [comment]
@DelusionalOptimist I'll get it done.

**Comment by rootxrishabh** [comment]
@DelusionalOptimist There are some build errors, fixing them might require dependency updates. If yes, should we upgrade dependencies across Kubearmor?

**Comment by rootxrishabh** [comment]
Kubearmor Controller uses configurations such as [MetricsBindAddress](https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.14/pkg/manager/manager.go#L225C2-L225C2) and [port](https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.14/pkg/manager/manager.go#L240) which were present in current package `controller-runtime:0.14.x` but have been removed in latest. Should we explore to reconfigure the controller or stick to old release?

**Comment by rootxrishabh** [comment]
libbpf-based [vulnerabilities](https://osv.dev/vulnerability/OSV-2021-1562) are still present as the patch is yet to be released across all releases.

**Comment by rootxrishabh** [comment]
> Kubearmor Controller uses configurations such as [MetricsBindAddress](https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.14/pkg/manager/manager.go#L225C2-L225C2) and [port](https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.14/pkg/manager/manager.go#L240) which were present in current package `controller-runtime:0.14.x` but have been removed in latest. Should we explore to reconfigure the controller or stick to old release?

Shifting to an older release makes sense as of now.

**Comment by rootxrishabh** [comment]
@nyrahul failing snyk tests require maintainers access, could you please take a look?

**Comment by DelusionalOptimist** [comment]
>> Kubearmor Controller uses configurations such as [MetricsBindAddress](https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.14/pkg/manager/manager.go?rgh-link-date=2024-01-11T22%3A41%3A24Z#L225C2-L225C2) and [port](https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.14/pkg/manager/manager.go?rgh-link-date=2024-01-11T22%3A41%3A24Z#L240) which were present in current package controller-runtime:0.14.x but have been removed in latest. Should we explore to reconfigure the controller or stick to old release?

> Shifting to an older release makes sense as of now.

Sure, we can handle these later as per need. :+1:

File: pkg/KubeArmorController/go.mod [hunk 1]
```diff
@@ -2,80 +2,72 @@ module github.com/kubearmor/KubeArmor/pkg/KubeArmorController
...
-	sigs.k8s.io/controller-runtime v0.14.6
-	sigs.k8s.io/yaml v1.3.0
+	github.com/go-logr/logr v1.4.1
+	github.com/onsi/ginkgo/v2 v2.13.0
+	github.com/onsi/gomega v1.29.0
+	k8s.io/api v0.29.0
+	k8s.io/apiextensions-apiserver v0.29.0
+	k8s.io/apimachinery v0.29.0
+	k8s.io/client-go v0.29.0
+	sigs.k8s.io/controller-runtime v0.15.1
```
Comment (DelusionalOptimist) [inline pkg/KubeArmorController/go.mod]: Is it possible to match at least the minor version of controller-runtime here with the Operator and KubeArmor. :eyes:

**Review by DelusionalOptimist — COMMENTED** [review]
All other updates look good. :+1:

**Comment by DelusionalOptimist** [comment]
> libbpf-based [vulnerabilities](https://osv.dev/vulnerability/OSV-2021-1562) are still present as the patch is yet to be released across all releases.

cc @daemon1024

**Comment by daemon1024** [comment]
I believe the libbpf vulnerability is due it existing as a submodule right?
I have created this issue to track that https://github.com/kubearmor/KubeArmor/issues/1580
Even if the fix is released, it won't be a simple update for us.

File: protobuf/go.mod [hunk 2]
```diff
@@ -17,8 +17,8 @@ require (
 )
 
 require (
-	golang.org/x/net v0.10.0 // indirect
-	golang.org/x/sys v0.8.0 // indirect
-	golang.org/x/text v0.9.0 // indirect
+	golang.org/x/net v0.17.0 // indirect
```
Comment (daemon1024) [inline protobuf/go.mod]: https://security.snyk.io/vuln/SNYK-GOLANG-GOLANGORGXCRYPTOSSH-6130669
```
golang.org/x/net@v0.17.0 › golang.org/x/crypto@v0.14.0 
```

Snyk failure cause

Comment (rootxrishabh) [reply protobuf/go.mod]: Upgraded to golang.org/x/net@v0.20.0 uses [golang.org/x/crypto v0.18.0](https://github.com/golang/net/blob/v0.20.0/go.mod#L6), still snyk fails. Am I missing something?

Comment (rootxrishabh) [reply protobuf/go.mod]: > Upgraded to golang.org/x/net@v0.20.0 uses [golang.org/x/crypto v0.18.0](https://github.com/golang/net/blob/v0.20.0/go.mod#L6), still snyk fails. Am I missing something?

Solved!
Could you provide a bit more insight on `security/snyk (nyrahul)` ?

Comment (Aryan-sharma11) [reply protobuf/go.mod]: https://security.snyk.io/vuln/SNYK-GOLANG-GITHUBCOMGOJOSEGOJOSE-6070736
Snyk failure cause

Comment (rootxrishabh) [reply protobuf/go.mod]: @Aryan-sharma11 I test all our go.mod files using snyk CLI. Other than `golang.org/x/net/http2@0.20.0`(low risk vuln, fix yet to implemented upstream) I can't find any vulnerability. Also, [github.com/go-jose/go-jose](https://security.snyk.io/vuln/SNYK-GOLANG-GITHUBCOMGOJOSEGOJOSE-6070736) is a transitive dependency to a package. Can we ignore this?

**Pull request closed without merging**